src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs

```diff
@@ -104,6 +153,37 @@ public void Initialize(bool skipDefaultAdapters)
         /// <inheritdoc/>
         public void DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
         {
+            if (this.proxyOperationManager == null)
+            {
+                // In case we have an active test session, we always prefer the already
+                // created proxies instead of the ones that need to be created on the spot.
+                this.proxyOperationManager = TestSessionPool.Instance.TakeProxy(
+                    this.testSessionInfo,
+                    discoveryCriteria.Sources.First(),
+                    runSettings);
+
+                if (this.proxyOperationManager == null)
+                {
+                    // If the proxy creation process based on test session info failed, then
+                    // we'll proceed with the normal creation process as if no test session
+                    // info was passed in in the first place.
+                    // 
+                    // WARNING: This should not normally happen and it raises questions
+                    // regarding the test session pool operation and consistency.
+                    EqtTrace.Warning("ProxyDiscoveryManager creation with test session failed.");
+
+                    this.proxyOperationManager = new ProxyOperationManager(
```

Comment:
Did you mean to ask if there's a way for the caller to check if testhost(s) rather than session(s) are going to be reused ? If so, read below please. If you actually meant session reuse, you can reuse the session as long as it is alive, and the session stays alive as long as TE needs it. You should be able to issue discovery/run, and then discovery/run requests again on the same test session, provided all session parameters stay the same.

There's no way for the caller to get that info now (i.e. if a testhost is going to be reused or not). We can potentially use the discovery/run events handler to let the caller know if the testhost is going to be reused because that's the only time we know for sure if we took out a testhost from the pool or decided to create another one on the spot.

Reply:
I'm just hoping that the scenarios where TP does this fallback are predictable and that it won't be common that TE will believe it's using sessions but in reality it won't

Reply:
Well, TE will probably use test sessions either way, but now that we are going to introduce the LRU cache we've been talking about lately the testhost reuse behavior won't be predictable anymore.